### PR TITLE
Defer hashing in staged ledger diff application

### PR DESCRIPTION
**Pull request by georgeee**
Problem: when performing transaction application, all merkle ledger hashes are recomputed for every account update. This is wasteful for a number of reasons:

  1. Transactions normally contain more than one account update and a hash computed for an account update is sometimes overwritten by a subsequent account update
  2. Ledger's depth is 35, whereas only around 2^18 records are actually populated. This means that each account update induces an overhead of at least 17 hashes (most of these wasteful)
  3. When an account is touched in a few transactions of the same block, it's gets hashed a few times, whereas in fact only the final hash is truly needed

Solution: defer computation of hashes in mask. When an account is added, it's stacked in a list `unhashed_accounts` of masks which is processed at the time of next access to hashes.

This fix improved performance of handling 9-account-update transactions by ~60% (measured by #14582 on top of #15979 and #15978).

Explain how you tested your changes:
* [x] measured performance with #14582

Checklist:

- [x] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? None

### CODE REVIEW

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 1]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+      |> List.iter ~f:(Tuple2.uncurry @@ self_set_hash_impl t)
+
+    let finalize_hashes t =
+      let unhashed_accounts = t.unhashed_accounts in
+      if not @@ List.is_empty unhashed_accounts then (
+        t.unhashed_accounts <- [] ;
+        finalize_hashes_do t unhashed_accounts )
+
+    (** Either copies accumulated or initializes it with the parent being used as the [base]. *)
+    let to_accumulated t =
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:383]: Note to myself: mostly unchanged, only `finalize_hashes t;` line is added now.

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 2]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+      | None, Ok base ->
+          { base
+          ; next = t.maps
+          ; current = t.maps
+          ; detached_next_signal = t.detached_parent_signal
+          }
+      | None, Error loc ->
+          raise (Dangling_parent_reference (t.uuid, loc))
+
+    let self_set_hash t address hash =
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:398]: Note to myself: unchanged except for the first line.

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 3]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+      finalize_hashes t ;
+      self_set_hash_impl t address hash
+
     let set_inner_hash_at_addr_exn t address hash =
       assert_is_attached t ;
       assert (Addr.depth address <= t.depth) ;
       self_set_hash t address hash
 
+    let hashes_and_ancestor t =
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:407]: Note to myself: mostly unchanged except for the first line. `maps_and_ancestor` was left as it is, and `hashes_and_ancestor` additionally runs `finalize_hashes` before the sub-call.

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 4]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+      List.stable_sort ~compare:(on_snd Location.compare) unhashed_accounts
+      |> List.remove_consecutive_duplicates ~which_to_keep:`First
+           ~equal:(on_snd Location.equal)
+      |> with_merkle_path_batch |> compute_merge_hashes
+      |> List.iter ~f:(Tuple2.uncurry @@ self_set_hash_impl t)
+
+    let finalize_hashes t =
+      let unhashed_accounts = t.unhashed_accounts in
+      if not @@ List.is_empty unhashed_accounts then (
+        t.unhashed_accounts <- [] ;
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:379]: Is it important to first nullify the existing `unhashed_accounts` before calling `finalize_hashes_do`?... Would the other way around work?

It's confusing because `finalize_hashes_do` also takes `t` as an argument, so you pass `unhashed_accounts` into it twice technically. `finalize_hashes_do` could just read/nullify these internally.

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 5]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+            (self_merkle_path ~current_location:t.current_location
+               ~depth:t.depth ~hashes )
+          ~fixup_path:(fixup_merkle_path ~hashes)
+          (List.map ~f:snd accs)
+        |> List.map2_exn accs ~f:(fun (a, loc) p ->
+               (a, Location.to_path_exn loc, p) )
+      in
+      let on_snd f (_, a) (_, b) = f a b in
+      List.stable_sort ~compare:(on_snd Location.compare) unhashed_accounts
+      |> List.remove_consecutive_duplicates ~which_to_keep:`First
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:371]: What's happening here? You seem to only keep the first account for each duplicated location. I assume you want to keep the last one instead, that is the last update that was pushed to the list? Or are we prepending to the list?

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 6]
```diff
@@ -81,6 +81,18 @@ module Make (Inputs : Inputs_intf.S) = struct
           *)
     }
 
+  (** Type for an account that wasn't yet hashed.
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:84]: General question regarding your comment from the PR description:

>When an account is touched in a few transactions of the same block, it's gets hashed a few times, whereas in fact only the final hash is truly needed

I'm wondering if this does not change the original application logic. Assume tx1 changes A and B, and tx2 changes C and A. So A1 becomes A2 and then A3, B1 becomes B2, C1 becomes C2. Surely, after tx1 and tx2 the final account state for A is A3, but don't we /at all/ care about the intermediate Merkle tree? Can we just rehash the MT based on A3 B2 C2? 

(I suspect yes, just double checking)

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 7]
```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
           acc.current <- f acc.current ;
           acc.next <- f acc.next )
 
-    let self_set_hash t address hash =
+    let self_set_hash_impl t address hash =
       update_maps t ~f:(fun maps ->
           { maps with hashes = Map.set maps.hashes ~key:address ~data:hash } )
 
+    let path_batch_impl ~fixup_path ~self_lookup ~base_lookup locations =
```
Comment (volhovm) [inline src/lib/merkle_mask/masking_merkle_tree.ml:216]: Note to myself: slightly rewritten to not have `t` as an argument, and to not pass `hashes` to the `self_lookup` parameter (it's passed/embedded into `self_lookup` on the caller level).

**Review by volhovm — COMMENTED** [review]
After spending >1.5h in total looking into this PR:
* It seems like a cool PR / idea to start with.
* I have no way of judging if this change is as you intended (unfortunately). This is a dense module describing a non-trivial data structure with close-to-zero documentation. On the positive side, I think I figured out the general sentiment of what you tried to achieve, which is already good.
* I'd like to see tests for this module
* I'd like to see explicit type annotations AND comments for most internal functions (think >5-10 lines of code; small helpers can be skipped), added in this PR and those moved around. Reading this for the first time trying to guess what the functions are doing without comments/types is _hard_.

(offtop rant: `open module T.S; module T = T.T; let type t = t; let type a x = t.a x; let rec impl a t = impl T.impl go a`. 2 step indentation (really hard to tell where code blocks end), 80 symbol wrapping, no explicit type annotations, almost no comments. This codebase is impossible to review without opening a whole IDE. I hope we're moving towards a more reader-friendly codestyle?.. :skull: )

**Comment by georgeee** [comment]
!ci-build-me

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 8]
```diff
@@ -81,6 +81,18 @@ module Make (Inputs : Inputs_intf.S) = struct
           *)
     }
 
+  (** Type for an account that wasn't yet hashed.
+  
+      Manual definitions of sexp are needed to avoid warning 4 in autogenerated code
+    *)
+  type unhashed_account_t = Account.t option * Location.t
+
+  let sexp_of_unhashed_account_t =
```
[the remark this answers is not in the document]
Comment (georgeee) [reply src/lib/merkle_mask/masking_merkle_tree.ml]: Well, they were used for derivation of `sexp` for some times. But then I realized we indeed do not need them, here is a PR: #16761

**Comment by georgeee** [comment]
!ci-build-me

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 9]
```diff
@@ -201,15 +194,206 @@ module Make (Inputs : Inputs_intf.S) = struct
...
+        match task with
+        | [] | [ (_, _, []) ] ->
+            acc'
+        | _ ->
+            impl acc' (height + 1) (converge height task)
+      in
+      let hash_account =
+        Option.value_map ~default:Hash.empty_account ~f:Hash.hash_account
+      in
+      Fn.compose (impl [] 0) (List.map ~f:(Tuple3.map_fst ~f:hash_account))
```
Comment (glyh) [inline src/lib/merkle_mask/masking_merkle_tree.ml:339]: If I'm understanding correctly, we're supposed to pass in a list of type
```
( Account.t option
           * Addr.t
           * [ `Left of Hash.t | `Right of Hash.t ] list )
           list
```
Where each adjacent item in this list should either be siblings, or the later being a parent of the previous. And we're resulted in a reversed, up-to-down chain from root to some path. This part make sense.

However, for adjacent pairs of leaves, we're passing in duplicate data and praying on the precondition to hold. I.e. they're indeed siblings. I suggest redesign this a bit so we only need to pass in one list of [`Left of Hash.t | `Right of Hash.t] instead of multiple, if possible.

In short, what's preventing us to pass in a "(Account.t option * Addr.t * [`Left Hash.t | `Right Hash.t]) list"?I get it that we might not be a full binary tree, but then we could just extend on the `Left, `Right cases?

Comment (glyh) [reply src/lib/merkle_mask/masking_merkle_tree.ml:339]: NVM I misunderstood this

File: src/lib/merkle_mask/masking_merkle_tree.ml (hunk 4 above)
Comment (georgeee) [reply src/lib/merkle_mask/masking_merkle_tree.ml:364]: Agree that passing `t` as an argument is a bit confusing.

This pattern of setting unhashed accounts to an empty list, and then passing the previous value as an argument is common in parallel programming, so that while processing the unhashed accounts list, we are sure that the list won't be growing.

Here it's not that important, of course, here there is no parallelism.

To make implementation clearer, I'd consider passing following arguments to `finalize_hashes_do`:
```ocaml
finalize_hashes_do unhashed_accounts
~path_lookup: (fun query -> let { hashes; _ }, ancestor = maps_and_ancestor t in
        path_batch_impl
          ~base_lookup:(Base.merkle_path_batch ancestor)
          ~self_lookup:
            (self_merkle_path ~current_location:t.current_location
               ~depth:t.depth ~hashes )
          ~fixup_path:(fixup_merkle_path ~hashes) query)) ~set_hash:(self_set_hash_impl t)
```

So that we'll remove the `t` parameter entirely, hence localize management of `t`'s state to `finalize_hashes`, leaving the interesting logic to `finalize_hashes_do` which was my original idea.

File: src/lib/merkle_mask/masking_merkle_tree.ml (hunk 6 above)
Comment (georgeee) [reply src/lib/merkle_mask/masking_merkle_tree.ml:78]: > Can we just rehash the MT based on A3 B2 C2?

We sometimes need intermediate hashes in between applications of `tx1` and `tx2`. What this PR actually does is that this will be an explicit ask of some code a few layers above. Rather than automatic recomputation of hashes each time an account is written to.

Also, in context of a single transaction, we never need intermediate hashes between application of individual account updates. I.e. if you are updating accounts `a1`, `a2`, `a3` in a single transaction, prior to this PR you'd recompute merkle tree root three times. Whereas in all of the cases we need no more than a single recomputation of root.
